executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,544 @@
+            offset: k,
+        }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Display> AffineSymbolicExpression<T, V> {
+    pub fn from_known_variable(var: V) -> Self {
+        SymbolicExpression::from_var(var).into()
+    }
+    pub fn from_unknown_variable(var: V) -> Self {
+        AffineSymbolicExpression {
+            coefficients: [(var, T::from(1).into())].into_iter().collect(),
+            offset: SymbolicExpression::from(T::from(0)),
+        }
+    }
+    pub fn from_number(n: T) -> Self {
+        SymbolicExpression::from(n).into()
+    }
+
+    /// If this expression does not contain unknown variables, returns the symbolic expression.
+    pub fn try_to_known(&self) -> Option<&SymbolicExpression<T, V>> {
+        if self.coefficients.is_empty() {
+            Some(&self.offset)
+        } else {
+            None
+        }
+    }
+
+    /// Tries to multiply this expression with another one.
+    /// Returns `None` if the result would be quadratic.
+    pub fn try_mul(&self, other: &Self) -> Option<Self> {
+        if !self.coefficients.is_empty() && !other.coefficients.is_empty() {
+            return None;
+        }
+        let (multiplier, coefficients, offset) = if self.coefficients.is_empty() {
+            (&self.offset, &other.coefficients, &other.offset)
+        } else {
+            (&other.offset, &self.coefficients, &self.offset)
+        };
+        let coefficients = coefficients
+            .iter()
+            .map(|(var, coeff)| (var.clone(), coeff * multiplier))
+            .collect();
+        let offset = offset * multiplier;
+        Some(AffineSymbolicExpression {
+            coefficients,
+            offset,
+        })
+    }
+
+    /// Solves the equation `self = 0` and returns how to compute the solution.
+    pub fn solve(
+        &self,
+        range_constraints: &impl RangeConstraintSet<V, T>,
+    ) -> Result<Vec<Effect<T, V>>, EvalError<T>> {
+        match self.coefficients.len() {
+            0 => {
+                return if self.offset.is_known_zero() {
+                    Ok(vec![])
+                } else {
+                    Err(EvalError::ConstraintUnsatisfiable(self.to_string()))
+                };
+            }
+            1 => {
+                let (var, coeff) = self.coefficients.iter().next().unwrap();
+                assert!(!coeff.is_known_zero());
+                let value = self.offset.field_div(&coeff.neg());
+                let assignment = Effect::Assignment(var.clone(), value);
+                if let Some(coeff) = coeff.try_to_number() {
+                    assert!(!coeff.is_zero(), "Zero coefficient has not been removed.");
+                    return Ok(vec![assignment]);
+                } else {
+                    return Ok(vec![
+                        Assertion::assert_is_nonzero(coeff.clone()),
+                        assignment,
+                    ]);
+                }
+            }
+            _ => {}
```

Comment:
Let's move the rest of the function inside here :) That way you can remove a bunch of `return` statements.